### PR TITLE
#162414285 Bookmark articles

### PR DESCRIPTION
**Pull request by Bobsar0**
#### What does this PR do?
Create, Get and Remove Bookmarks on an Article
#### Description of Task to be completed?
 • Create Bookmark controller and routes
 • Create controller methods (`createOrRemove`, `read`) for performing actions on bookmark resource
 • Modify related model/migration files and associations
 • Write tests
 • Create a helper function to handle check for resource existence
 • Refactor bookmark controller using helper function

#### How should this be manually tested?
 • Clone this branch
 • CD into the repo
 • Run `npm install`
 • Run `npm run migrate && npm run seed`
 • Run `npm run dev`
 • Use postman to check the following endpoints:
```
 • http://localhost/3000/articles/:slug/bookmark (POST OR DELETE);
`Initial request creates bookmark, the next request removes the bookmark..and so on`
 • http://localhost/3000/articles/:slug/bookmarks (GET article's bookmarks);
 • http://localhost/3000/users/:userName/bookmarks (GET user's bookmarks); 
```

#### Background contexts?
Note that you must be an authenticated user to perform endpoint operations.
 • Login with the following credentials to access endpoints
 ``` 
 email: steve@now.com
 password: Blahblah
``` 
 • Use generated token as `bearer token` auth type to access endpoints.

#### What are the relevant pivotal tracker stories?
[Create, Get and Remove Article Bookmarks](https://www.pivotaltracker.com/story/show/162414285)

### CODE REVIEW

File: server/controllers/BookmarkController.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+   * @returns {object} api route response
+   */
+  static async create(req, res) {
+    try {
+      const { slug } = req.params;
+      const { userId } = req.user;
+      const article = await Util.resourceExists(Article, { slug });
+      if (!article) return response(res, 404, 'failure', 'Article not found');
+
+      Bookmark.findOrCreate({
```
Comment (jesseinit) [inline server/controllers/BookmarkController.js]: Hi Steve,
Good work on the PR
Any reason why you didn't use sync await on here?

File: server/routes/api/bookmark.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { Router } from 'express';
+import BookmarkController from '../../controllers/BookmarkController';
+import AuthMiddleware from '../../middlewares/AuthMiddleware';
+
+const bookmarkRoutes = Router();
+const path = '/articles/:slug/bookmark';
+
+bookmarkRoutes.post(path, AuthMiddleware.checkIfUserIsAuthenticated, BookmarkController.create);
```
Comment (jesseinit) [inline server/routes/api/bookmark.js]: Route definition doesn't seem to be in-line with how other routes have been defined and agreed by the team. Any reason for this decision?

**Comment by tejiri4** [comment]
LGTM, yet.

**Comment by Makwe-O** [comment]
Apart from the way you defined the routes, This PR looks fine. LGTM Yet

File: server/controllers/BookmarkController.js (hunk 1 above)
Comment (rafmme) [reply server/controllers/BookmarkController.js]: @Bobsar0 just a reminder, don't forget to close this conversation when resolved

**Comment by rafmme** [comment]
LGTM, yet

File: server/controllers/BookmarkController.js (hunk 1 above)
Comment (Bobsar0) [reply server/controllers/BookmarkController.js]: Thanks Jesse. Will resolve

File: server/routes/api/bookmark.js (hunk 2 above)
Comment (Bobsar0) [reply server/routes/api/bookmark.js]: Yeah. Just following DRY principles. Will change though. Thanks for noticing

**Comment by ebzeal** [comment]
LGTM yet

**Pull request merged**